### PR TITLE
[Misc] NFC: Fix random build warnings

### PR DESCRIPTION
**Pull request by davezarzycki**
Unused variables/methods, language extensions, extra semicolons, intentional self assignment, platform specific quirks, etc. Also downright horror shows. For example, I'm not sure that anybody verified that swift-reflection-test.c actually works.

### CODE REVIEW

**Comment by davezarzycki** [comment]
@swift-ci please smoke test

**Review by gottesmm — COMMENTED** [review]
Quick question about the "horror show".

File: lib/Frontend/ArgsToFrontendOutputsConverter.cpp [hunk 1]
```diff
@@ -133,7 +133,7 @@ OutputFilesComputer::OutputFilesComputer(
     const FrontendOptions::ActionType requestedAction,
     const llvm::opt::Arg *moduleNameArg, const StringRef suffix,
     const bool hasTextualOutput)
-    : Args(args), Diags(diags), InputsAndOutputs(inputsAndOutputs),
```
Comment (jrose-apple) [inline lib/Frontend/ArgsToFrontendOutputsConverter.cpp]: Shouldn't the argument be dropped too? cc @davidungar

File: include/swift/SIL/SILInstruction.h [hunk 2]
```diff
@@ -1222,36 +1222,37 @@ struct SILDebugVariable {
...
       /// variable name present. This typically only happens for
       /// instructions that were created from parsing SIL assembler.
-      unsigned NameLength : 14;
+      int_type NameLength : 14;
+      /// The source function argument position from left to right
+      /// starting with 1 or 0 if this is a local variable.
+      int_type ArgNo : 16;
     } Data;
-  };
+  } Bits;
```
Comment (jrose-apple) [inline include/swift/SIL/SILInstruction.h]: Why this change?

**Comment by davezarzycki** [comment]
Hi @gottesmm – Thanks for the feedback. Done.

@swift-ci please smoke test

File: unittests/runtime/Metadata.cpp [hunk 3]
```diff
@@ -297,12 +297,12 @@ ProtocolDescriptor ProtocolNoWitnessTable{
...
 
 TEST(MetadataTest, getExistentialMetadata) {
-  const ProtocolDescriptor *protoList1[] = {};
+  const ProtocolDescriptor *const proto = nullptr;
   RaceTest_ExpectEqual<const ExistentialTypeMetadata *>(
     [&]() -> const ExistentialTypeMetadata * {
       auto any = swift_getExistentialTypeMetadata(ProtocolClassConstraint::Any,
                                                   /*superclass=*/nullptr,
-                                                  0, protoList1);
+                                                  0, &proto);
```
Comment (jrose-apple) [inline unittests/runtime/Metadata.cpp]: Why this change?

File: include/swift/SIL/SILInstruction.h (hunk 2 above)
Comment (davezarzycki) [reply include/swift/SIL/SILInstruction.h]: Because the following is generated by the default warnings with clang/LLVM top-of-tree:
```
include/swift/SIL/SILInstruction.h:1227:5: warning: anonymous types declared in an anonymous union are an extension [-Wnested-anon-types]
    struct {
    ^
```

File: unittests/runtime/Metadata.cpp (hunk 3 above)
Comment (davezarzycki) [reply unittests/runtime/Metadata.cpp]: Because the following is generated by the default warnings with clang/LLVM top-of-tree:
```
unittests/runtime/Metadata.cpp:300:44: warning: zero size arrays are an extension [-Wzero-length-array]
  const ProtocolDescriptor *protoList1[] = {};
                                           ^
```

File: lib/Frontend/ArgsToFrontendOutputsConverter.cpp (hunk 1 above)
Comment (davezarzycki) [reply lib/Frontend/ArgsToFrontendOutputsConverter.cpp]: Probably. I just wanted to quiet the unused field warning.

File: unittests/runtime/Metadata.cpp (hunk 3 above)
Comment (davezarzycki) [reply unittests/runtime/Metadata.cpp]: I briefly looked into that, but I lost interest in verifying that all of the downstream APIs could handle the array itself being null. In theory, none of this matters. Everything should be checking the "num protos" parameter first.

**Comment by davezarzycki** [comment]
Just to be clear here, I'm only trying to quiet trivial warnings with conservative fixes. I'm not trying to take responsibility for how any particular subsystem should work or interpret any particular dead code as an awkward todo list.

File: lib/Frontend/ArgsToFrontendOutputsConverter.cpp (hunk 1 above)
Comment (davidungar) [reply lib/Frontend/ArgsToFrontendOutputsConverter.cpp]: Yes, please by all means drop the argument, too. (Tx @jrose-apple )

**Comment by davezarzycki** [comment]
Added @davidungar's feedback.

@swift-ci please smoke test

File: unittests/runtime/Metadata.cpp (hunk 3 above)
Comment (jrose-apple) [reply unittests/runtime/Metadata.cpp]: That's exactly why I'd pass null here. If it doesn't cause any issues, it's a better test, and if it *does* cause issues you found a bug (which you aren't responsible for fixing).

Comment (davezarzycki) [reply unittests/runtime/Metadata.cpp]: Well, 10975 tests passed on my Linux box, so let's go with `nullptr`. :-)

**Comment by davezarzycki** [comment]
@swift-ci please smoke test
